src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,126 @@
+  private final IDrawable anvil, slot;
+  private static final int WIDTH = 134;
+  private static final int HEIGHT = 66;
+  private static final int X_OFFSET = -6;
+  private static final int Y_OFFSET = -15;
+  private static final int SLOT_SIZE = 18;
+  private static final int ITEM_SIZE = 16;
+
+  public ToolBuildingCategory(IGuiHelper guiHelper) {
+    this.icon = guiHelper.createDrawableItemStack(TinkerTools.pickaxe.get().getRenderTool());
+    this.background = guiHelper.createDrawable(BACKGROUND_LOC, 122, 77, WIDTH, HEIGHT);
+    this.slot = guiHelper.createDrawable(BACKGROUND_LOC, 24, 14, SLOT_SIZE, SLOT_SIZE);
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<List<ItemStack>> partsAndExtras = Stream.concat(recipe.getAllToolParts().stream(),
+      recipe.getExtraRequirements().stream().map(ingredient -> Arrays.asList(ingredient.getItems()))).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    int missingSlots = partsAndExtras.size() - layoutSlots.size();
+
+    // check layout slots for issues
+    if (missingSlots > 0) {
+      TConstruct.LOG.error(String.format("Tool part count is greater than layout slot count for %s!", recipe.getId()));
+      layoutSlots = new ArrayList<>(layoutSlots);
+      for (int additionalSlot = 0; additionalSlot < missingSlots; additionalSlot++) {
+        layoutSlots.add(new LayoutSlot(null, null, additionalSlot * SLOT_SIZE -X_OFFSET, -Y_OFFSET, null));
+      }
+    } else if (missingSlots < 0) {
+      partsAndExtras = new ArrayList<>(partsAndExtras);
+      for (int additionalItem = 0; additionalItem > missingSlots; additionalItem--){
+        partsAndExtras.add(List.of(ItemStack.EMPTY));
+      }
+    }
+
+    for (int i = 0; i < layoutSlots.size(); i++) {
+      builder.addSlot(RecipeIngredientRole.INPUT, layoutSlots.get(i).getX() + X_OFFSET, layoutSlots.get(i).getY() + Y_OFFSET)
+             .addItemStacks(partsAndExtras.get(i));
+    }
+
+    ItemStack outputStack = recipe.getOutput() instanceof IModifiableDisplay modifiable ? modifiable.getRenderTool() : recipe.getOutput().asItem().getDefaultInstance();
+      builder.addSlot(RecipeIngredientRole.OUTPUT, WIDTH - 26, 23)
+        .addItemStack(outputStack);
+  }
+
+  @Override
+  public void draw(ToolBuildingRecipe recipe, IRecipeSlotsView recipeSlotsView, PoseStack stack, double mouseX, double mouseY) {
+    if (recipe.requiresAnvil()) {
+      this.anvil.draw(stack, 76, 44);
+    }
+
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    int missingSlots = recipe.getAllToolParts().size() + recipe.getExtraRequirements().size() - layoutSlots.size();
+
+    // check layout slots for issues
+    if (missingSlots > 0) {
+      TConstruct.LOG.error(String.format("Tool part count is greater than layout slot count for %s!", recipe.getId()));
```

Comment:
Hopefully the last thing, isn't this error also printing in `setRecipe`? Little redundant to print in both, especial since IIRC draw runs every frame so this message will be printing a lot. Just keep the `setRecipe` copy.